Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm

```diff
@@ -0,0 +1,121 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "ProgressBarMac.h"
+
+#if PLATFORM(MAC)
+
+#import "LocalDefaultSystemAppearance.h"
+#import "ProgressBarPart.h"
+
+namespace WebCore {
+
+ProgressBarMac::ProgressBarMac(ProgressBarPart& owningPart, ControlFactoryMac& controlFactory)
+    : ControlMac(owningPart, controlFactory)
+{
+}
+
+IntSize ProgressBarMac::cellSize(NSControlSize controlSize, const ControlStyle&) const
+{
+    static const std::array<IntSize, 4> sizes =
+    {
+        IntSize(0, 20),
+        IntSize(0, 12),
+        IntSize(0, 12),
+        IntSize(0, 20)
+    };
+    return sizes[controlSize];
+}
+
+IntOutsets ProgressBarMac::cellOutsets(NSControlSize controlSize, const ControlStyle&) const
+{
+    static const IntOutsets cellOutsets[] = {
+        // top right bottom left
+        { 0, 0, 1, 0 },
+        { 0, 0, 1, 0 },
+        { 0, 0, 1, 0 },
+        { 0, 0, 1, 0 },
+    };
+    return cellOutsets[controlSize];
+}
+
+void ProgressBarMac::draw(GraphicsContext& context, const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style)
+{
+    LocalDefaultSystemAppearance localAppearance(style.states.contains(ControlStyle::State::DarkAppearance), style.accentColor);
+
+    IntRect inflatedRect = IntRect(rect);
+    auto imageBuffer = context.createImageBuffer(inflatedRect.size(), deviceScaleFactor);
+    if (!imageBuffer)
+        return;
+
+    ContextContainer cgContextContainer(imageBuffer->context());
+    CGContextRef cgContext = cgContextContainer.context();
+
+    auto& progressBarPart = owningProgressBarPart();
+    auto controlSize = calculateControlSize(style);
+    bool isIndeterminate = progressBarPart.position() < 0;
+    bool isActive = true; // isActive(renderObject)
```

Comment:
This should read a bit from `ControlStyle`.

Reply:
Fixed. We should be using
```
bool isActive = style.states.contains(ControlStyle::State::WindowActive);
```